File: installer/installer_utils/platforms/mac.py
import os
from .base import InstallConfig, PlatformHandler
from ..utils import check_dependency_installed, run_command


class MacHandler(PlatformHandler):
    def install_dependencies(self, config: InstallConfig):
        if not check_dependency_installed("which brew"):
            raise RuntimeError("Homebrew not found, please install it first.")
        run_command("brew install git python")

    def setup_autostart(self, config: InstallConfig):
        plist_dir = os.path.expanduser("~/Library/LaunchAgents")
        os.makedirs(plist_dir, exist_ok=True)
        plist_path = os.path.join(plist_dir, "com.octane.startup.plist")
        with open(plist_path, "w") as f:
            f.write(
                f"""<?xml version="1.0" encoding="UTF-8"?>
                <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
                <plist version="1.0">
                <dict>
                    <key>Label</key>
                    <string>com.octane.startup</string>
                    <key>ProgramArguments</key>
                    <array>
                        <string>{os.path.join(config.install_path, '.venv/bin/python')}</string>
                        <string>{os.path.join(config.install_path, 'main.py')}</string>
                    </array>
                    <key>RunAtLoad</key>
                    <true/>
                </dict>
                </plist>"""
            )
            
    def get_activate_cmd(self, config: InstallConfig):
        return f"source {config.install_path}/.venv/bin/activate"

    def get_create_env_cmd(self, config: InstallConfig):
        return f"python3 -m venv {config.install_path}/.venv"

    def get_python_cmd(self):
        return "python3"

    def get_env_file(self):
        return ".env-example-unix"

File: source/Octane-Manager/installer_utils/platforms_specific/mac.ts
import * as fs from 'fs';
import * as path from 'path';
import { EnvironmentUtil } from '../modules/environment_util';
import { InstallConfig } from '../config';
import { DependencyInstaller } from '../modules/dependency_installer';
import { InstallerBase } from '../modules/installer_base';
import { check_dependency_installed, run_command } from '../utils';

class MacInstallerBase extends InstallerBase {
    setup_autostart(config: InstallConfig): void {
        const plist_dir = path.join(os.homedir(), "Library", "LaunchAgents");
        fs.mkdirSync(plist_dir, { recursive: true });
        const plist_path = path.join(plist_dir, "com.octane.startup.plist");
        const plist_content = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.octane.startup</string>
    <key>ProgramArguments</key>
    <array>
        <string>${path.join(config.install_path, '.venv/bin/python')}</string>
        <string>${path.join(config.install_path, 'main.py')}</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>`;
        fs.writeFileSync(plist_path, plist_content);
    }
}

class MacDependencyInstaller extends DependencyInstaller {
    install_system_dependencies(config: InstallConfig): void {
        const missing_dependencies = ["git", "python3"].filter(cmd => !check_dependency_installed(`which ${cmd}`));

        if (missing_dependencies.length > 0) {
            if (!check_dependency_installed("which brew")) {
                throw new Error(
                    `The following dependencies are missing: ${missing_dependencies.join(', ')}, either install them or install brew and run the installer again.`
                );
            }
            run_command("brew install git python");
        }
    }
}

class MacEnvironmentUtil extends EnvironmentUtil {
    // Additional macOS-specific environment setup can be added here
}
